apollo-ast/src/main/kotlin/com/apollographql/apollo3/ast/internal/ExecutableValidationScope.kt

```diff
@@ -310,6 +320,111 @@ internal class ExecutableValidationScope(
+    selectionSet.validateDeferDirectives("")
+  }
+
+  /**
+   * If a label is passed to a `@defer` directive, it must not be a variable, and it must be unique within all other `@defer` directives in
+   * the document.
+   *
+   * Also ensure that any `@defer` directive found when walking fragments on an operation have a unique path + label (Apollo-specific
+   * validation).
+   *
+   * For instance: this is invalid:
+   * ```
+   * query Query1 {
+   *   computers {
+   *   ...ComputerFields @defer  # path is computer
+   *   ...ComputerFields @defer  # path is computer
+   *   }
+   * }
+   * ```
+   *
+   * Also invalid:
+   * ```
+   * query Query2 {
+   *   computers {
+   *     ...ComputerFields
+   *     ...ComputerFields2
+   *     }
+   *   }
+   *
+   *   fragment ComputerFields on Computer {
+   *     screen {
+   *       ...ScreenFields @defer  # path is computer.screen
+   *     }
+   *   }
+   *
+   *   fragment ComputerFields2 on Computer {
+   *     screen {
+   *       ...ScreenFields @defer  # path is computer.screen
+   *     }
+   * }
+   * ```
+   */
+  private fun GQLSelectionSet.validateDeferDirectives(path: String) {
+    for (selection in selections) {
+      when (selection) {
+        is GQLField -> {
+          val fieldPath = if (path.isEmpty()) selection.name else "$path.${selection.name}"
+          selection.selectionSet?.validateDeferDirectives(fieldPath)
+        }
+        is GQLInlineFragment -> {
+          val deferDirective = selection.children.firstOrNull { it is GQLDirective && it.name == "defer" } as GQLDirective?
+          if (deferDirective != null && deferDirective.validateDeferDirective(path, deferDirective.sourceLocation)) return
+          selection.selectionSet.validateDeferDirectives(path)
+        }
+        is GQLFragmentSpread -> {
+          val deferDirective = selection.children.firstOrNull { it.name == "defer" }
+          if (deferDirective != null && deferDirective.validateDeferDirective(path, deferDirective.sourceLocation)) return
+          fragmentDefinitions[selection.name]!!.selectionSet.validateDeferDirectives(path)
+        }
+      }
+    }
+  }
+
+  /**
+   * @return true if an issue was found, false otherwise
+   */
+  private fun GQLDirective.validateDeferDirective(path: String, sourceLocation: SourceLocation): Boolean {
+    val label = arguments?.arguments?.firstOrNull { it.name == "label" }?.value
+    if (label is GQLVariableValue) {
+      registerIssue(
+          message = "@defer label argument must not be a variable",
+          sourceLocation = sourceLocation
+      )
+      return true
+    }
+
+    var labelStringValue = ""
+    if (label != null) {
+      labelStringValue = (label as GQLStringValue).value
+      if (labelStringValue in deferDirectiveLabels) {
+        registerIssue(
+            message = "@defer label '$labelStringValue' must be unique within all other @defer directives in the document",
+            sourceLocation = sourceLocation
```

Comment:
It could be nice to "remember" where the other `@defer` sourceLocation is to give more actionable errors